client/gui/InterfaceObjectConfigurable.cpp

```diff
@@ -0,0 +1,386 @@
+{
+	auto font = readFont(config["font"]);
+	auto alignment = readTextAlignment(config["alignment"]);
+	auto color = readColor(config["color"]);
+	auto group = std::make_shared<CLabelGroup>(font, alignment, color);
+	if(!config["items"].isNull())
+	{
+		for(const auto & item : config["items"].Vector())
+		{
+			auto position = readPosition(item["position"]);
+			auto text = readText(item["text"]);
+			group->add(position.x, position.y, text);
+		}
+	}
+	return group;
+}
+
+std::shared_ptr<CSlider> InterfaceObjectConfigurable::buildSlider(const JsonNode & config) const
+{
+	auto position = readPosition(config["position"]);
+	int length = config["size"].Integer();
+	auto style = config["style"].String() == "brown" ? CSlider::BROWN : CSlider::BLUE;
+	auto itemsVisible = config["itemsVisible"].Integer();
+	auto itemsTotal = config["itemsTotal"].Integer();
+	auto value = config["selected"].Integer();
+	bool horizontal = config["orientation"].String() == "horizontal";
+	return std::make_shared<CSlider>(position, length, callbacks.at(config["callback"].String()), itemsVisible, itemsTotal, value, horizontal, style);
+}
+
+std::shared_ptr<CAnimImage> InterfaceObjectConfigurable::buildImage(const JsonNode & config) const
+{
+	auto position = readPosition(config["position"]);
+	auto image = config["image"].String();
+	int group = config["group"].isNull() ? 0 : config["group"].Integer();
+	int frame = config["frame"].isNull() ? 0 : config["frame"].Integer();
+	return std::make_shared<CAnimImage>(image, frame, group, position.x, position.y);
+}
+
+std::shared_ptr<CFilledTexture> InterfaceObjectConfigurable::buildTexture(const JsonNode & config) const
+{
+	auto image = config["image"].String();
+	auto rect = readRect(config["rect"]);
+	return std::make_shared<CFilledTexture>(image, rect);
+}
+
+std::shared_ptr<CShowableAnim> InterfaceObjectConfigurable::buildAnimation(const JsonNode & config) const
+{
+	auto position = readPosition(config["position"]);
+	auto image = config["image"].String();
+	ui8 flags = 0;
+	if(!config["repeat"].Bool())
+		flags |= CShowableAnim::EFlags::PLAY_ONCE;
+	
+	int group = config["group"].isNull() ? 0 : config["group"].Integer();
+	auto anim = std::make_shared<CShowableAnim>(position.x, position.y, image, flags, 4, group);
+	if(!config["alpha"].isNull())
+		anim->setAlpha(config["alpha"].Integer());
+	if(!config["callback"].isNull())
+		anim->callback = std::bind(callbacks.at(config["callback"].String()), 0);
+	if(!config["frames"].isNull())
+	{
+		auto b = config["frames"]["start"].Integer();
+		auto e = config["frames"]["end"].Integer();
+		anim->set(group, b, e);
+	}
+	return anim;
+}
+
+std::shared_ptr<CIntObject> InterfaceObjectConfigurable::buildWidget(JsonNode config) const
+{
+	assert(!config.isNull());
+	//overrides from variables
+	for(auto & item : config["overrides"].Struct())
+	{
+		config[item.first] = variables[item.second.String()];
+	}
+	
+	auto type = config["type"].String();
+	if(type == "picture")
```

Comment:
Not sure about this if-forest.
Perhaps generate static std::map<std::string, std::function> instead?

Reply:
Or, create method `protected: registerWidget(std::string name, std::function builder)`
that will internally add name & builder functor to std::map and any inheritors can register their own widgets instead of relying on `buildCustomWidget` method

Reply:
Change like this seems overkilling, buildCustomWidget doesn't work in a way designed:
`No viable conversion from 'std::shared_ptr<CIntObject> (TemplatesDropBox::*)(const JsonNode &)' to 'std::function<std::shared_ptr<CIntObject> (const InterfaceObjectConfigurable *, const JsonNode &)>'`

Reply:
Static map won't work here - disregard my first comment, it needs to be non-static member of InterfaceObjectConfigurable
```
using BuilderFunctor = std::function<std::shared_ptr<CIntObject>(const JsonNode &)>;
std::map<std::string, builderFunctor> functors;

void InterfaceObjectConfigurable::registerWidget(std::string name, const BuilderFunctor & builder)
{
    functors["button"] = builder;
}

InterfaceObjectConfigurable::InterfaceObjectConfigurable()
{
    registerWidget("button", [this](const JsonNode & node){
        return buildButton(node);
    }
}

TemplatesDropBox::TemplatesDropBox()
{
    registerWidget("custom", [this](const JsonNode & node){
        return buildCustomWidget(node);
    }
}
```

Reply:
If you still can't make it work - then OK, leave it as it.

Reply:
Unfortunately, for some of those builders JsonNode is not enough - I've added pointer to InterfaceObjectConfigurable (this), because builder may use internal variables from class instance.
Still can be solved with std::bind, but that's my concern - seems like overcomplicating as amount of types are limited

Reply:
One option would be to replace if-forest of basic classes with std::map but keep call to buildCustomWIdget as it is.

Anyway, don't see this as blocker - this is minor change that can be done at any point later, taking into account any future changes.

Reply:
ok, looks like I found a more or less nice way of doing that